wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerHandler.java

```diff
@@ -99,23 +107,76 @@ protected boolean shouldCopyHeader(@Nullable String workspaceId) {
     AtomicReference<Connection> connectionRef = new AtomicReference<>();
     try {
       getContext().execute(datasetContext -> connectionRef.set(store.get(connectionId)));
     } catch (TransactionFailureException e) {
       if (e.getCause() instanceof ConnectionNotFoundException) {
-        ServiceUtils.notFound(responder, e.getCause().getMessage());
-        return null;
+        throw (ConnectionNotFoundException) e.getCause();
       }
-      ServiceUtils.error(responder, e.getMessage());
-      return null;
+      throw e;
     }
     Connection connection = connectionRef.get();
+    if (connection.getType() == null) {
+      throw new BadRequestException("Connection type must be specified.");
+    }
     if (expectedType != connection.getType()) {
-      error(responder, "Invalid connection type set, this endpoint only accepts GCS connection type");
-      return null;
+      throw new BadRequestException(String.format("Expected connection type '%s' but found '%s'.",
+                                                  expectedType, connection.getType()));
     }
     return connection;
   }
+
+  /**
+   * Utility method for executing an endpoint with common error handling and namespace checks built in.
+   * A response will always be sent after this method is called so the http responder should not be used after this.
+   * The endpoint logic should also not use the responder in any way.
+   *
+   * If the callable throws a {@link StatusCodeException}, the exception's status code and message will be used
+   * to create the response.
+   * If a {@link JsonSyntaxException} is thrown, a 400 response will be sent.
+   * If anything else if thrown, a 500 response will be sent.
+   * If nothing is thrown, the result of the callable will be sent as json.
+   *
+   * @param request the http request
+   * @param responder the http responder
+   * @param namespace the namespace to check for, or null if no check needs to be performed
+   * @param callable the endpoint logic to run
+   */
+  protected <T> void respond(HttpServiceRequest request, HttpServiceResponder responder, @Nullable String namespace,
+                             Callable<T> callable) {
+    if (namespace != null) {
+      try {
+        if (!getContext().getAdmin().namespaceExists(namespace)) {
+          responder.sendJson(HttpURLConnection.HTTP_NOT_FOUND,
+                             new ServiceResponse<Void>(String.format("Namespace '%s' does not exist", namespace)));
+          return;
+        }
+      } catch (IOException e) {
+        responder.sendJson(HttpURLConnection.HTTP_INTERNAL_ERROR, new ServiceResponse<Void>(e.getMessage()));
+        return;
+      }
+    }
+
+    try {
+      T results = callable.call();
+      responder.sendJson(results);
+    } catch (StatusCodeException e) {
+      responder.sendJson(e.getCode(), new ServiceResponse<>(e.getMessage()));
+    } catch (JsonSyntaxException e) {
```

Comment:
should we also handle NotFoundException and AlreadyExistsException?

Reply:
these extend StatusCodeException so they will be handled